### PR TITLE
MeanSegmentEncoderTransform

### LINKED ISSUE #233: Add `MeanSegmentEncoderTransform`
### 🚀 Feature Request

Add class `MeanSegmentEncoderTransform` that makes mean encoding of the segment.
It should make regressor with name `regressor_segment_mean`, that contains expanding mean of mean target across the segment.

### Motivation

This class can make predictions better.

### Proposal

Add class `MeanSegmentEncoderTransform` that makes mean encoding of the segment.
It should make regressor with name `regressor_segment_mean`, that contains expanding mean of mean target across the segment.

What is expanding mean encoding. It is encoding of the categorical variable based on mean value of the target. At each timestamp we calculate mean value among all previous timestamps -- it can save us from target leakage. For the test data mean on all train data is used.

This technique with some adjustments is used in CatBoost to work with categorical variables. You can read about it [here](https://towardsdatascience.com/how-catboost-encodes-categorical-variables-3866fb2ae640).

The main difficulty in realization of this task is to understand how to make this distinction between train and test data. For example, remembering timestamps on fit stage can be used.

Unfortunately, we can't add multiple permutations, as it violates the temporal structure. However, we can add quantization and prior like in CatBoost in the future if this method proves it successful.

### Test cases

1. Test that we have expanding mean on train, but overall mean on test.
2. Check that this encoding gives a big advantage to `LinearMultiSegmentModel`, that has only this feature. Try to predict something simple like almost constant segments.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alex-hse-repository**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna-ts/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna-ts/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Proposed Changes

Add MeanSegmentEncoderTransform
## Closing issues

closes #233

### CODE REVIEW

File: tests/test_transforms/test_segment_encoder_transform.py [hunk 1]
```diff
@@ -28,3 +33,77 @@ def test_dummy(dummy_df):
...
+
+    train, test = almost_constant_ts.train_test_split(test_size=horizon)
+    train.fit_transform([encoder])
+    model.fit(train)
+    future = train.make_future(horizon)
+    pred_mean_segment_encoding = model.forecast(future)
+
+    metric = MAE(mode="macro")
+
+    assert metric(pred_mean_segment_encoding, test) < 1
```
Comment (Mr-Geekman) [inline tests/test_transforms/test_segment_encoder_transform.py:109]: Try to use R2 here, it is more intuitive in sense of good/bad result. May be noise should be smaller.

File: tests/test_transforms/test_segment_encoder_transform.py [hunk 2]
```diff
@@ -28,3 +33,80 @@ def test_dummy(dummy_df):
...
+
+    train, test = almost_constant_ts.train_test_split(test_size=horizon)
+    train.fit_transform([encoder])
+    model.fit(train)
+    future = train.make_future(horizon)
+    pred_mean_segment_encoding = model.forecast(future)
+
+    metric = R2(mode="macro")
+
+    # R2=0 => model predict the optimal constant
```
Comment (alex-hse-repository) [inline tests/test_transforms/test_segment_encoder_transform.py:111]:
```suggestion
    # R2=0 => model predicts the optimal constant
```

**Review by Mr-Geekman — APPROVED** [review]
